include/global/state.hpp

```diff
@@ -513,38 +549,69 @@ namespace NP {
 			void update_job_finish_times(const Schedule_state& from,
 				Job_index j, Interval<Time> start_times,
 				Interval<Time> finish_times,
-				const std::vector<Job_index>& jobs_with_pending_succ)
+				const std::vector<Job_with_pending_succ>& jobs_with_pending_succ)
 			{
 				Time lst = start_times.max();
 				Time lft = finish_times.max();
 
+				bool single_core = (core_avail.size() == 1);
+
 				job_finish_times.reserve(jobs_with_pending_succ.size());
 
 				auto it = from.job_finish_times.begin();
-				for (Job_index job : jobs_with_pending_succ)
+				for (const Job_with_pending_succ& job : jobs_with_pending_succ)
 				{
-					if (job == j)
-						job_finish_times.emplace_back(job, finish_times);
+					if (job.job_idx == j)
+						job_finish_times.emplace_back(job.job_idx, finish_times, job.is_certainly_finished);
 					else {
 						// we find the finish time interval of `job` from the previous state. 
 						// Note that if `job` has non-completed successors in the new state,
 						// it must have had non-completed successors in the previous state too, 
 						// thus there is no risk to reach the end iterator
-						while (it->first != job) {
+						while (it->job_idx != job.job_idx) {
 							assert(it != from.job_finish_times.end());
 							it++;
 						}
-						Time job_eft = it->second.min();
-						Time job_lft = it->second.max();
+						Time job_eft = it->finish_time.min();
+						Time job_lft = it->finish_time.max();
 						// if there is a single core, then we know that 
 						// jobs that were disptached in the past cannot have 
 						// finished later than when our new job starts executing
-						if (core_avail.size() == 1)
+						if (single_core)
 						{
 							if (job_lft > lst)
 								job_lft = lst;
 						}
-						job_finish_times.emplace_back(job, Interval<Time>{ job_eft, job_lft });
+
+						// Assuming:
+						// (1) j_low cannot start until at least 1 core is available.
+						// (2) So if all cores are certainly occupied until j_pred is finished, we can disregard j_pred.
+						//
+						// We will prove the following claim: 
+						// (4) If ft(j).min() <= ca(1).min && ft(j).max() <= ca(2).min() then no core can be available
```

Comment:
If you don't want to keep assumption (3), it's probably better to rename (4) to (3). It looks a bit weird to have (1) followed by (2) followed by (4).